specs/kyber/src/serialize.rs

```diff
@@ -1,108 +1,203 @@
 ///   lossy compression takes place)
-/// - `|bits_per_coefficient| < |parameters::BITS_PER_COEFFICIENT|`, since
-///   otherwise when `deserialize_little_endian` operates on 12 bits at a time,
+/// - `bits_per_coefficient < BITS_PER_COEFFICIENT`, since
+///   otherwise when `byte_decode` operates on 12 bits at a time,
 ///   it is not injective: the values 3329 + 1 and 1 for example both fit into
 ///   12 bits and map to the same `KyberFieldElement`
 ///
-/// Otherwise `deserialize_little_endian` is not injective and therefore has
-/// no left inverse.
-pub fn serialize_little_endian(
-    re: KyberPolynomialRingElement,
-    bits_per_coefficient: usize,
-) -> Vec<u8> {
-    let mut serialized: Vec<u8> = Vec::new();
-
-    for bit_vector in to_little_endian_bit_vector(re, bits_per_coefficient).chunks(8) {
-        let mut byte_value: u8 = 0;
-        for (i, &bit) in bit_vector.iter().enumerate() {
-            byte_value |= bit << i;
-        }
+/// Otherwise `byte_decode` is not injective and therefore has no left inverse.
+///
+/// N.B.: This function asserts that `bits_per_coefficient <= 12`
+///
+/// This function implements <strong>Algorithm 4</strong> of the NIST FIPS 203 standard, which is
+/// reproduced below:
+///
+/// ```plaintext
+/// Input: integer array F ∈ ℤₘ²⁵⁶, where m = 2ᵈ if d < 12 and m = q if d = 12.
+/// Output: byte array B ∈ 𝔹^{32d}.
+/// for(i ← 0; i < 256; i++)
+///     a ← F[i]
+///     for(j ← 0; j < d; j++)
+///         b[i·d + j] ← a mod 2
+///         a ← (a − b[i·d + j])/2
+///     end for
+/// B ← BitsToBytes(b)
+/// return B
+/// ```
+///
+/// The NIST FIPS 203 standard can be found at
+/// <https://csrc.nist.gov/pubs/fips/203/ipd>.
+pub fn byte_encode(bits_per_coefficient: usize, re: KyberPolynomialRingElement) -> Vec<u8> {
+    assert!(bits_per_coefficient <= BITS_PER_COEFFICIENT);
 
-        serialized.push(byte_value);
+    let mut re_bits = BitVector::new();
+
+    // for(i ← 0; i < 256; i++)
+    for coefficient in re.coefficients() {
+        // a ← F[i]
+        let mut coefficient_value = coefficient.value;
+
+        // for(j ← 0; j < d; j++)
+        for _ in 0..bits_per_coefficient {
+            let bit = coefficient_value % 2;
+
+            // b[i·d + j] ← a mod 2
+            re_bits.push(bit as u8);
+
+            // a ← (a − b[i·d + j])/2
+            coefficient_value = (coefficient_value - bit) / 2;
+        }
     }
 
-    serialized
+    bits_to_bytes(re_bits)
 }
 
-fn field_element_from_little_endian_bit_vector(bit_vector: BitVector) -> KyberFieldElement {
+fn field_element_from_bits(bits: BitVector) -> KyberFieldElement {
+    assert!(bits.len() <= BITS_PER_COEFFICIENT);
+
+    let modulus = 2u16.pow(bits.len().as_u32());
     let mut value: u16 = 0;
-    for (i, bit) in bit_vector.into_iter().enumerate() {
-        value |= u16::from(bit) << i;
+
+    for (i, bit) in bits.into_iter().enumerate() {
+        value += ((bit as u16) * 2u16.pow(i.as_u32())) % modulus;
     }
```

Comment:
How about this? For better readability.

```suggestion
        value += ((bit as u16) * (1 << i)) % modulus;
```

Reply:
Changed.